extensions/bit-manipulation-1/README.md

```diff
@@ -0,0 +1,65 @@
+# General design
+
+**Extension State: Under Development**  
+**Requires: Base, VWI, CP2.0**  
+**CPUID Bit: CP2.4**
+
+# Overview
+
+This extension adds rotate with carry and various bit operations.
+
+# Added Instructions
+
+These instructions are in the expanded calculation opcode section of instructions.
+
+### Opcode table
+
+| `C CCCC CCCC` | NAME      | Operation                                              | Flags | Comment     |
+|---------------|-----------|--------------------------------------------------------|-------|-------------|
+| `0 0000 1000` | `RCL`     | <code>C:A ← (C:A << B) &#124; (C:A >> -B)</code>       | `ZNC` | (1) (2) (3) |
+| `0 0000 1001` | `RCR`     | <code>A:C ← (A:C >> B) &#124; (A:C << -B)</code>       | `ZNC` | (1) (2) (3) |
+| `0 0000 1010` | `POPCNT`  | <code>A ← POPCNT(B)</code>                             | `ZN`  | (4)         |
+| `0 0000 1011` | `GREV`    | <code>A ← GREV(A, B)</code>                            | `ZN`  | (5)         |
+| `0 0000 1100` | `CTZ`     | <code>A ← CTZ(B)</code>                                | `ZN`  | (6)         |
+| `0 0000 1101` | `CLZ`     | <code>A ← CLZ(B)</code>                                | `ZN`  | (7)         |
+| `0 0000 1110` | `NOT`     | <code>A ← ~B</code>                                    | `ZN`  |             |
+| `0 0000 1111` | `ANDN`    | <code>A ← ~A &#38; B</code>                            | `ZN`  |             |
+| `0 0001 1000` | `LSB`     | <code>A ← B &#38; -B</code>                            | `ZN`  |             |
+| `0 0001 1001` | `LSMSK`   | <code>A ← B ^ (B - 1)</code>                           | `ZN`  |             |
+| `0 0001 1010` | `RLSB`    | <code>A ← B &#38; (B - 1)</code>                       | `ZN`  |             |
```

Comment:
Let's include a description of these last 4 operations. I don't think we should expect the average person reading this document to already be familiar with what these expressions do.

A page on the wiki motivating these instructions might be a good idea as well? Honestly it's a bit difficult to motivate `rcl`/`rcr` for shift amounts other than 1. Do we know of any use-cases for that? It's a little bit awkward, but I might not be opposed to only specifying support for rcl/rcr by 1 with a feature or possibly later extension specifying support for other amounts.

EDIT: I did some research on this. Amounts other than 1 incur slowdowns on x86 and according to StackOverflow are "rarely used, if ever."

Reply:
I have no problems with making rcr and rcl limited to shifts by 1. We could also just remove rcl in that case because of the adc overlap

Reply:
If we turn `rcr`/`rcl` into unary operations, it is once again similar to the XOR/NOT operation, so I wouldn't be opposed to keeping it. For symmetry with everything else in the assembly we could have the pseudo instructions `rcr, %rx0, 1` which are only valid with immediate=1.

Reply:
We could also make it so that `rcr, a, b` performs the operation on `b` and stores it in `a`

Reply:
There are other options, like the original idea of leaving values other than 1 to a feature or something.

I can't even figure out if ARM has rotate through carry instructions. Most old microcontrollers don't seem to support values other than 1. x86 added support for that via microcoding at some point but I can't find any use cases of it. I think I actually link Ender's idea here. The only downside is if using these instructions to implement a bigint or softfloat library, I think you might actually want the operation `rcr [a], 1` to be one instruction, which isn't possible for a version that isn't in-place. That's a small cost though and I'm not even sure it applies.

Reply:
In that case, I'm fine with fixing the second argument at 1 unless a feature is implemented if we don't want to do my above suggestion.

Reply:
It looks like ARM has `rcr dst, src` which does rotate right with carry by 1. It's called `rrx` (rotate right with extend). They don't have `rcl`, but that's because they just use `adc dst, src, src` to emulate it.
https://developer.arm.com/documentation/dui0489/i/arm-and-thumb-instructions/rrx

Reply:
OK. I'm down to replace the instructions specified here with src->dst rotates by 1. There's a clear distinction between such an `rcl` and an `adc` so it can stay. The immediate modes seem completely useless but it would be unnecessary complication to reserve them, I think. And, technically, `rcl r, 15` would be a 1-instruction (3-byte) way to put 30 in a register without FI. Not that that sounds useful.